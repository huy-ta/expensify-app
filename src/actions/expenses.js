import uuid from 'uuid';
import database from '../firebase/firebase';

const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
});

const startAddExpense = (expenseData = {}) => { // after integrating redux-thunk, this can work
    return (dispatch) => {
        const {
            description = '', 
            note = '', 
            amount = 0, 
            createdAt = 0 
        } = expenseData; // destructuring
        const expense = { description, note, amount, createdAt };
        return database.ref('expenses').push(expense).then((ref) => {
            dispatch(addExpense({
                id: ref.key,
                ...expense
            }));
        });
    };
};

const removeExpense = ({ id } = {}) => ({
    type: 'REMOVE_EXPENSE',
    id
});

const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
});

export { addExpense, removeExpense, editExpense, startAddExpense };