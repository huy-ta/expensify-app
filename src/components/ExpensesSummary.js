import React from 'react';
import { connect } from 'react-redux';
import selectExpenses from '../selectors/expenses';
import selectExpensesTotal from '../selectors/expenses-total';
import numeral from 'numeral';

export const ExpensesSummary = (props) => {
    const expensesCount = props.expenses.length;
    const expensesTotal = expensesCount === 0 ? 0 : selectExpensesTotal(props.expenses);
    return (
        <div>
            <h2>Viewing {expensesCount} expense(s) totalling {numeral(expensesTotal / 100).format('$0,0.00')}</h2>
        </div>
    );
};

const mapStateToProps = (state) => ({
    expenses: selectExpenses(state.expenses, state.filters)
});

export default connect(mapStateToProps)(ExpensesSummary);