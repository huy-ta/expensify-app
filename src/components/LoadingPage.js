import React from 'react';
import LoadingGif from '../assets/images/loader.gif';

const LoadingPage = () => (
    <div className="loader">
        <img className="loader__image" src={LoadingGif} />
    </div>
);

export default LoadingPage;