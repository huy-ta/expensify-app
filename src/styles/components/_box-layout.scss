$bg-img: url('../assets/images/bg.jpg');

.box-layout {
    align-items: center;
    background-image: linear-gradient(rgba(0, 0, 0, 0.5), rgba(0, 0, 0, 0.5)), $bg-img;
    background-repeat: no-repeat;
    background-size: cover;
    display: flex;
    height: 100vh;
    justify-content: center;
    width: 100vw;
}

.box-layout__box {
    background: fade-out(white, .1);
    border-radius: 3px;
    padding: $l-size $l-size;
    text-align: center;
    width: 25rem;
}

.box-layout__title {
    margin: 0 0 $m-size 0;
    line-height: 1;
}

.box-layout__button {
    align-items: center;
    display: flex;
    justify-content: center;
    width: 100%;
}

.box-layout__icon {
    display: inline-block;
    margin-right: $s-size;
}