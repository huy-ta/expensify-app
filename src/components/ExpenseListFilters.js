import React from 'react';
import { connect } from 'react-redux';
import { setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate } from '../actions/filters';
import { DateRangePicker } from 'react-dates';

export class ExpenseListFilters extends React.Component {
    state = {
        calendarFocused: null
    };

    onTextChange = (e) => {
        this.props.setTextFilter(e.target.value);
    };

    onSelectChange = (e) => {
        if (e.target.value === "date") {
            this.props.sortByDate();
        } else if (e.target.value === "amount") {
            this.props.sortByAmount(); 
        }
    }

    onDatesChange = ({ startDate, endDate }) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    };

    onFocusChange = (calendarFocused) => {
        this.setState(() => ({ calendarFocused }));
    };

    render() {
        return (
            <div>
                <input 
                    type="text" 
                    value={this.props.filters.text} 
                    onChange={this.onTextChange}
                />
                <select 
                    value={this.props.filters.sortBy} 
                    onChange={this.onSelectChange}
                >
                    <option value="date">Date</option>
                    <option value="amount">Amount</option>
                </select>
                <DateRangePicker 
                    startDate={this.props.filters.startDate}
                    startDateId="hehe"
                    endDate={this.props.filters.endDate}
                    endDateId="hoho"
                    onDatesChange={this.onDatesChange}
                    focusedInput={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    showClearDates={true}
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                />
            </div>
        );
    }
}

const mapStateToProps = (state) => ({
    filters: state.filters
});

const mapDispatchToProps = (dispatch) => ({
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    sortByDate: () => dispatch(sortByDate()),
    sortByAmount: () => dispatch(sortByAmount()),
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate))
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);