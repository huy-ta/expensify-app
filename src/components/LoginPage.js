import React from 'react';
import { connect } from 'react-redux';
import { startLogin } from '../actions/auth';
import googleIcon from '../assets/images/google.svg';

export const LoginPage = ({ startLogin }) => (
    <div className="box-layout">
    <div className="box-layout__box">
        <h1 className="box-layout__title">Expensify</h1>
        <p>It's time to get your expenses under control.</p>
        <div>
            <button className="box-layout__button" onClick={startLogin}>
                <img className="box-layout__icon" src={googleIcon} width='32' />
                Login with Google
            </button>
        </div>
    </div>
    </div>
);

const mapDispatchToProps = (dispatch) => ({
    startLogin: () => dispatch(startLogin())
});

export default connect(undefined, mapDispatchToProps)(LoginPage);