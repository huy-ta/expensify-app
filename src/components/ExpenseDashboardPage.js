import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';
import myImg from '../assets/images/46035.jpg';

const ExpenseDashboardPage = () => (
    <div>
        <img src={myImg} width='auto'/>
        <ExpenseListFilters />
        <ExpenseList />
    </div>
);

export default ExpenseDashboardPage;