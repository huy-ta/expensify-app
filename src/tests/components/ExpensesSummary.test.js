import React from 'react';
import { shallow } from 'enzyme';
import { ExpensesSummary } from '../../components/ExpensesSummary';
import expenses from '../fixtures/expenses';

describe('ExpensesSummary', () => {
    test('should render ExpensesSummary correctly with no expense', () => {
        const wrapper = shallow(<ExpensesSummary expenses={[]} />);
        expect(wrapper).toMatchSnapshot();
    });
});

describe('ExpensesSummary', () => {
    test('should render ExpensesSummary correctly with one expense', () => {
        const wrapper = shallow(<ExpensesSummary expenses={[expenses[0]]} />);
        expect(wrapper).toMatchSnapshot();
    });
});

describe('ExpensesSummary', () => {
    test('should render ExpensesSummary correctly with multiple expenses', () => {
        const wrapper = shallow(<ExpensesSummary expenses={expenses} />);
        expect(wrapper).toMatchSnapshot();
    });
});