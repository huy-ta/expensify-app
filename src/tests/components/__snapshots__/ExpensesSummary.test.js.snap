// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExpensesSummary should render ExpensesSummary correctly with multiple expenses 1`] = `
<div>
  <h2>
    Viewing 
    3
     expense(s) totalling 
    $1,141.95
  </h2>
</div>
`;

exports[`ExpensesSummary should render ExpensesSummary correctly with no expense 1`] = `
<div>
  <h2>
    Viewing 
    0
     expense(s) totalling 
    $0.00
  </h2>
</div>
`;

exports[`ExpensesSummary should render ExpensesSummary correctly with one expense 1`] = `
<div>
  <h2>
    Viewing 
    1
     expense(s) totalling 
    $1.95
  </h2>
</div>
`;
